tests/tests/src/untagged_variants_heuristic.mjs

```diff
@@ -0,0 +1,22 @@
+// Generated by ReScript, PLEASE EDIT WITH CARE
+
+
+function f(x) {
+  if (x !== null) {
+    return;
+  }
+  console.log("abc");
+}
+
+function ff(x) {
+  if (x === "G" || x === "F" || x === "E" || x === "D" || x === "C" || x === "B" || x === "A") {
```

Comment:
@cknitt see this case that is not worse

This is why I propose a heuristic based on how many cases with, and without payloads exist.
To determine which code to generate.

Reply:
@cometkim do you have a suggestion on how to set the threshold to maximise things like perf/code size etc?
There are these 2 ways of generating code. One is clearly better with a lot of cases without payload (`ff` here). The other is better when there are very few cases without payload and many cases with payload (e.g. json). The question is how to decide in the middle.

Reply:
> @cknitt see this case that is now worse
> 
> This is why I propose a heuristic based on how many cases with, and without payloads exist. To determine which code to generate.

Yes, sounds good to me, a heuristic-based approach will definitely improve the current situation without making things worse like in the example here. 👍

Reply:
I think this improves the predictability because the output is easier to understand. Untagged variants are not generally expected to be the best performance.

However, from a perf perspective, `typeof` checks are always slightly better than full literal checks, especially if it allows you to skip other branches.

I value how much more understandable it is rather than the actual code size. If the number of circuits is the same or similar, there is no real difference. It's just that `typeof` checks tend to be harder to read and understand when they get complex.

> There are these 2 ways of generating code. One is clearly better with a lot of cases without payload (ff here). The other is better when there are very few cases without payload and many cases with payload (e.g. json). The question is how to decide in the middle.

Not sure if possible, I'd guess it would be nice to use the current on the default strategy but do the opposite if the check matches the `_` pattern.